File: Cisco/GitHubList/Python/1_1WordSearch.py
class Solution:
    def exist(self, board: list[list[str]], word: str) -> bool:
        M, N, W = len(board), len(board[0]), len(word)
        DIRECTIONS = [
            (0, 1),  # Right
            (0, -1),  # Left
            (-1, 0),  # Up
            (1, 0),  # Down
        ]

        def dfs(i: int, j: int, target: int) -> bool:  # O(4**L) time, O(L) space
            if target == W:
                return True

            if i >= M or i < 0 or j >= N or j < 0 or board[i][j] != word[target]:
                return False

            temp, board[i][j] = board[i][j], "#"  # Mark as visited

            for di, dj in DIRECTIONS:
                if dfs(i + di, j + dj, target + 1):
                    return True

            board[i][j] = temp
            return False

        # Word can begin from anywhere.
        for i in range(M):  # O(M)
            for j in range(N):  # O(N)
                if word[0] == board[i][j]:
                    if dfs(i, j, 0):
                        return True
        return False

File: Cisco/GitHubList/C++/1_1WordSearch.cpp
#include <vector>
#include <string>

using namespace std;

class Solution
{
    const vector<pair<int, int>> DIRECTIONS = {
        {0, 1}, {0, -1}, {1, 0}, {-1, 0}};

    const char VISITED = '#';

    bool dfs(vector<vector<char>> &board, const string &word, int i, int j, int k)
    {
        if (k == word.size())
            return true;

        if (i < 0 || i >= board.size() || j < 0 || j >= board[0].size() || board[i][j] != word[k])
            return false;

        char temp = board[i][j];
        board[i][j] = VISITED;

        for (const auto &[dx, dy] : DIRECTIONS)
        {
            int x = i + dx, y = j + dy;
            if (dfs(board, word, x, y, k + 1))
                return true;
        }

        board[i][j] = temp;
        return false;
    }

public:
    bool exist(vector<vector<char>> &board, string word)
    {
        for (int i = 0; i < board.size(); ++i)
            for (int j = 0; j < board[0].size(); ++j)
                if (board[i][j] == word[0] && dfs(board, word, i, j, 0))
                    return true;
        return false;
    }
};
